src/libstore/build/outer-derivation-goal.hh

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include "parsed-derivations.hh"
+#include "lock.hh"
+#include "store-api.hh"
+#include "pathlocks.hh"
+#include "goal.hh"
+
+namespace nix {
+
+struct DerivationGoal;
+
+struct OuterDerivationGoal : public Goal
+{
+    /**
+     * How to obtain a store path of the derivation to build.
+     */
+    ref<SingleDerivedPath> drvReq;
+
+    /**
+     * The path of the derivation, once obtained.
+     **/
+    std::optional<StorePath> optDrvPath;
+
+    /**
+     * The goal for the corresponding concrete derivation.
+     **/
+    std::shared_ptr<DerivationGoal> concreteDrvGoal;
+
+    /**
+     * The specific outputs that we need to build.
+     */
+    OutputsSpec wantedOutputs;
+
+    typedef void (OuterDerivationGoal::*GoalState)();
+    GoalState state;
+
+    /**
+     * The final output paths of the build.
+     *
+     * - For input-addressed derivations, always the precomputed paths
+     *
+     * - For content-addressed derivations, calcuated from whatever the
+     *   hash ends up being. (Note that fixed outputs derivations that
+     *   produce the "wrong" output still install that data under its
+     *   true content-address.)
+     */
+    OutputPathMap finalOutputs;
+
+    BuildMode buildMode;
+
+    OuterDerivationGoal(ref<SingleDerivedPath> drvReq,
+        const OutputsSpec & wantedOutputs, Worker & worker,
+        BuildMode buildMode = bmNormal);
+    virtual ~OuterDerivationGoal();
+
+    void timedOut(Error && ex) override;
+
+    std::string key() override;
+
+    void work() override;
+
+    /**
+     * Add wanted outputs to an already existing derivation goal.
+     */
+    void addWantedOutputs(const OutputsSpec & outputs);
+
+    /**
+     * The states.
+     */
+    void getDerivation();
+    void loadAndBuildDerivation();
+    void buildDone();
+
+    JobCategory jobCategory() override { return JobCategory::Build; };
```

Comment:
What if it's substitutable?

It's unclear from the `jobCategory` docs what it's for or how it should behave, because it doesn't have docs.

Reply:
I think this for max jobs stuff. This goal does no work until itself so hmmm perhaps it should be another category.

Reply:
- https://github.com/NixOS/nix/pull/8833

Reply:
Still seems that this needs work. The goal shouldn't squat a build slot while it's realising the `.drv`. It seems that that would instead be something for the goal that actually builds or substitutes.

Maybe it `jobCategory()` should be `optional<JobCategory>` or `data ResourceTaken = LocalBuild | Substitution | None` or whatever.

We can't just make the return value state dependent either, because that just messes up the counters. Non-deterministically even, so that'd be fun ;)

(Or maybe I'm getting it all wrong because I haven't worked much with this code yet)